### ISSUE TITLE
Network object as child exception on loadglobalscenes() 

### ISSUE THREAD

**Issue by FullsailStudent1337** [FirstGearGames/FishNet#42]
1. Fishnet by default when using the Spawn function drops the instantiated object on the root of the scene by default.
2. There are situations when this is undesirable so we would need to call gameobject.transform.SetParent(ParentGO.transform)
3. When Unloading a scene Fishnet Cycles through ServerManager.Objects.Spawned.Values in order to generate a list of all spawned objects during the function call Scenemanager.MoveClientHostObjects()
4. **This is where things go wrong** MoveClientHostObjects() calls 
UnitySceneManager.MoveGameObjectToScene(nob.gameObject, moveScene);
 which according to Unity Documentation: "You can only move root GameObjects from one Scene to another. This means the GameObject to move must not be a child of any other GameObject in its Scene. This only works on GameObjects being moved to a Scene that is already loaded (additive)."

TLDR  if a Spawn()'d network object is not in the root and is visible to the client and host this exception is triggered.   I am not sure if you can do this, but in order to resolve this I assume you could do a loop to find the NOB's root level parent and pass that object instead to the moveScene.   Tested with  FN 1.4.5  and 1.4.7  Unity version: 2021.3.1f1

Relevant code:
Instantiation script:

```
public void Start()
    {
        
        if(InstanceFinder.IsServer)
        {
            InstanceFinder.SceneManager.OnClientPresenceChangeEnd += ServerPanelSpawnDataListener;
        }
    }
    [Server]
    private void ServerPanelSpawnDataListener(ClientPresenceChangeEventArgs obj)
    {
        Debug.Log("ServerPanelSpawnDataListener Fired");
        ServerRequestPanelSpawnData(obj.Connection);
    }
    [TargetRpc]
    private void ServerRequestPanelSpawnData(NetworkConnection clientNC)
    {
        Debug.Log("ServerRequestPanelSpawnData fired");
        SpawnPlayerDataFrame(GameManager.instance.playerName);
    }
    [ServerRpc(RequireOwnership = false)]
    public void SpawnPlayerDataFrame(string playername, NetworkConnection conn = null)
    {
        Debug.Log("SpawnPlayerDataFrame() fired");
        NetworkObject PlayerPanel = Instantiate(PlayerDataPrefab);
        PlayerPanel.GetComponent<PlayerPanelLogic>()._playerName = playername;


        InstanceFinder.ServerManager.Spawn(PlayerPanel, Owner);
    }
  
```
Instantiated object reassignment of parent:
  ```

public void Start()
    {
      
        this.transform.SetParent(GameObject.FindGameObjectWithTag("LobbyPlayersWidget").transform);
        this.transform.localScale = new Vector3(1, 1, 1);
    }
  
```
  Request Scene Change:
 ```

      public void LoadScene()
    {
        Debug.Log("loading scene" + Scene_Name);
        SceneLoadData sld = new SceneLoadData(Scene_Name);
        sld.ReplaceScenes = ReplaceOption.All;
        InstanceFinder.SceneManager.LoadGlobalScenes(sld);
        
    }
 
```

**Comment by FirstGearGames** [comment]
This is intended behavior. As you've discovered Unity does not allow moving child objects through scenes. Currently nested nobs are supported but you cannot unnest them at runtime. There's several reasons for this and a design decision has to be discussed before it can be changed. There's basically no one size fits all solution.

If you want to move an object it needs to be instantiated and set as root (meaning it can be under a game object holder prior but not nested).

In the near future objects will spawn with the same parent on clients. This is already mostly complete.

**Comment by FullsailStudent1337** [comment]
Thanks for the timely response,  I have implemented some logic that seems to resolve this issue for my needs:
        ```
private void MoveClientHostObjects(Scene scene, bool asServer)
{
            /* The asServer isn't really needed. I could only call
             * this method when asServer is true. But for the sake
             * of preventing user-error (me being the user this time)
             * I've included it into the parameters. */
            if (!asServer)
                return;
            //Don't need to perform if not host.
            if (!_networkManager.IsClient)
                return;

            NetworkConnection clientConn = _networkManager.ClientManager.Connection;
            /* It would be nice to see if the client wasn't even in the scene
             * here using SceneConnections but it's possible that the scene had been
             * wiped from SceneConnections earlier depending on how scenes are
             * loaded or unloaded. Instead we must iterate through spawned objects. */

            ListCache<NetworkObject> movingNobs = ListCaches.GetNetworkObjectCache();
            /* Rather than a get all networkobjects in scene
             * let's iterate the spawned objects instead. I imagine
             * in most scenarios iterating spawned would be faster.
             * That's a long one! */
            foreach (NetworkObject nob in _networkManager.ServerManager.Objects.Spawned.Values)
            {
                //Not in the scene being destroyed.
                if (nob.gameObject.scene != scene)
                    continue;
                //ClientHost doesn't have visibility.
                if (!nob.Observers.Contains(clientConn))
                    continue;

                /* If here nob is in the same being
                 * destroyed and clientHost has visiblity. */
                movingNobs.AddValue(nob);
            }

            int count = movingNobs.Written;
            if (count > 0)
            {
                Scene moveScene = GetDelayedDestroyScene();
                List<NetworkObject> collection = movingNobs.Collection;

                for (int i = 0; i < count; i++)
                {
                    NetworkObject nob = collection[i];
                    /* Force as not a scene object
                     * so that it becomes destroyed
                     * rather than disabled. */
                    nob.ClearRuntimeSceneObject();
                    /* If the object is already being despawned then
                     *just disable and move it. Otherwise despawn it
                    * on the server then move it. */
                    //Not deinitializing, despawn it then.
                    if (!nob.Deinitializing)
                        nob.Despawn();
                    else
                        nob.gameObject.SetActive(false);
                    //My custom Fix for child nob
                    if (nob.transform.parent)
                    {
                        Debug.Log("Found Nob Parent, moving up Heiarchy");
                        GameObject tempGO = nob.transform.parent.gameObject;
                        //find root parent object
                        while (tempGO.transform.parent)
                        {
                            Debug.Log("Found GameObject Parent, moving up Heiarchy");
                            tempGO = tempGO.transform.parent.gameObject;
                        }
                        UnitySceneManager.MoveGameObjectToScene(tempGO.gameObject, moveScene);
                    }    
                    else
                    UnitySceneManager.MoveGameObjectToScene(nob.gameObject, moveScene);
                }
            }
            ListCaches.StoreCache(movingNobs);
        }
```

Any glaring issue that you see here that could be detrimental?

**Comment by FirstGearGames** [comment]
I won't be able to put this into the main repo. If there were other nobs nested in the parent it would cause problems.
